database/19_Slow_Query/task_5/README.md

```diff
@@ -0,0 +1,148 @@
+# クイズ
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+<details>
+<summary>Table of Contents</summary>
+
+- [&#035;1 クイズ](#1-%E3%82%AF%E3%82%A4%E3%82%BA)
+- [&#035;2 クイズ](#2-%E3%82%AF%E3%82%A4%E3%82%BA)
+
+</details>
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## #1 クイズ
+
+実務ではスロークエリを直接確認するよりも、Percona社が提供している `pt-query-digest` のようなツールを使用してローカルのログファイルを確認したり、クラウド上でアプリを稼働させている場合には `Datadog` を使用してクエリの解析を行ったりすることが多いらしい。
+
+では実際に Percona社が提供している `pt-query-digest` をインストールして、スロークエリログを表示させてみましょう。
+
```

Comment:
以下の手順で導入してみました！

```bash
apt update
apt install -y wget
# dpkgインストールは自動で依存関係を解決してくれないので、以下を実施
apt install -y libdbi-perl libdbd-mysql-perl libio-socket-ssl-perl
# 実施したらエラーが出てインストールが失敗したので、fix
apt --fix-broken install
dpkg -i percona-toolkit_3.3.0-1.stretch_amd64.deb
```

で実行してみました！

```bash
# pt-query-digest 0771bb71246a-slow.log

# 180ms user time, 20ms system time, 16.00k rss, 11.00k vsz
# Current date: Fri Apr 16 11:50:33 2021
# Hostname: 0771bb71246a
# Files: 0771bb71246a-slow.log
# Overall: 9 total, 4 unique, 0.00 QPS, 0.01x concurrency ________________
# Time range: 2021-04-15T11:28:32 to 2021-04-15T12:25:59
# Attribute          total     min     max     avg     95%  stddev  median
# ============     ======= ======= ======= ======= ======= ======= =======
# Exec time            18s   212ms      4s      2s      4s      2s      3s
# Lock time            5ms   114us     3ms   508us     3ms   990us   131us
# Rows sent          2.64M       9 650.41k 300.15k 619.64k 146.45k 283.86k
# Rows examine      12.95M 292.99k   2.71M   1.44M   2.62M   1.11M 915.49k
# Query size           420      23     113   46.67  112.70   26.05   46.83

# Profile
# Rank Query ID                            Response time Calls R/Call V/M
# ==== =================================== ============= ===== ====== ====
#    1 0x3858CDBCFFE772CDE369FCAE71AFA31D  14.1820 77.9%     4 3.5455  0.04 SELECT salaries
#    2 0x9CFA9370319A0E1E01C119E92F1928F5   2.8928 15.9%     1 2.8928  0.00 SELECT employees salaries
#    3 0xDB534C97AF86024F633F5856DF702BD1   0.6384  3.5%     3 0.2128  0.00 SELECT employees
# MISC 0xMISC                               0.4968  2.7%     1 0.4968   0.0 <1 ITEMS>

# Query 1: 0.00 QPS, 0.01x concurrency, ID 0x3858CDBCFFE772CDE369FCAE71AFA31D at byte 409
# This item is included in the report because it matches --limit.
# Scores: V/M = 0.04
# Time range: 2021-04-15T11:35:36 to 2021-04-15T11:52:31
# Attribute    pct   total     min     max     avg     95%  stddev  median
# ============ === ======= ======= ======= ======= ======= ======= =======
# Count         44       4
# Exec time     77     14s      3s      4s      4s      4s   366ms      4s
# Lock time     13   636us   130us   207us   159us   204us    31us   181us
# Rows sent     43   1.14M 292.99k 292.99k 292.99k 292.99k       0 292.99k
# Rows examine  83  10.85M   2.71M   2.71M   2.71M   2.71M       0   2.71M
# Query size    45     192      48      48      48      48       0      48
# String:
# Databases    employees
# Hosts        localhost
# Users        root
# Query_time distribution
#   1us
#  10us
# 100us
#   1ms
#  10ms
# 100ms
#    1s  ################################################################
#  10s+
# Tables
#    SHOW TABLE STATUS FROM `employees` LIKE 'salaries'\G
#    SHOW CREATE TABLE `employees`.`salaries`\G
# EXPLAIN /*!50100 PARTITIONS*/
select sum(salary) from salaries group by emp_no\G

# Query 2: 0 QPS, 0x concurrency, ID 0x9CFA9370319A0E1E01C119E92F1928F5 at byte 2085
# This item is included in the report because it matches --limit.
# Scores: V/M = 0.00
# Time range: all events occurred at 2021-04-15T12:25:59
# Attribute    pct   total     min     max     avg     95%  stddev  median
# ============ === ======= ======= ======= ======= ======= ======= =======
# Count         11       1
# Exec time     15      3s      3s      3s      3s      3s       0      3s
# Lock time     74     3ms     3ms     3ms     3ms     3ms       0     3ms
# Rows sent     24 650.41k 650.41k 650.41k 650.41k 650.41k       0 650.41k
# Rows examine   7 943.40k 943.40k 943.40k 943.40k 943.40k       0 943.40k
# Query size    26     113     113     113     113     113       0     113
# String:
# Databases    employees
# Hosts        localhost
# Users        root
# Query_time distribution
#   1us
#  10us
# 100us
#   1ms
#  10ms
# 100ms
#    1s  ################################################################
#  10s+
# Tables
#    SHOW TABLE STATUS FROM `employees` LIKE 'employees'\G
#    SHOW CREATE TABLE `employees`.`employees`\G
#    SHOW TABLE STATUS FROM `employees` LIKE 'salaries'\G
#    SHOW CREATE TABLE `employees`.`salaries`\G
# EXPLAIN /*!50100 PARTITIONS*/
SELECT * FROM employees e JOIN salaries s ON e.emp_no = s.emp_no WHERE gender = "M" AND birth_date > "1960-01-01"\G

# Query 3: 0.00 QPS, 0.00x concurrency, ID 0xDB534C97AF86024F633F5856DF702BD1 at byte 0
# This item is included in the report because it matches --limit.
# Scores: V/M = 0.00
# Time range: 2021-04-15T11:28:32 to 2021-04-15T11:53:02
# Attribute    pct   total     min     max     avg     95%  stddev  median
# ============ === ======= ======= ======= ======= ======= ======= =======
# Count         33       3
# Exec time      3   638ms   212ms   214ms   213ms   208ms       0   208ms
# Lock time      8   394us   114us   150us   131us   144us    12us   125us
# Rows sent     32 878.98k 292.99k 292.99k 292.99k 292.99k       0 292.99k
# Rows examine   6 878.98k 292.99k 292.99k 292.99k 292.99k       0 292.99k
# Query size    16      69      23      23      23      23       0      23
# String:
# Databases    employees
# Hosts        localhost
# Users        root
# Query_time distribution
#   1us
#  10us
# 100us
#   1ms
#  10ms
# 100ms  ################################################################
#    1s
#  10s+
# Tables
#    SHOW TABLE STATUS FROM `employees` LIKE 'employees'\G
#    SHOW CREATE TABLE `employees`.`employees`\G
# EXPLAIN /*!50100 PARTITIONS*/
select * from employees\G
```

Reply:
![image](https://user-images.githubusercontent.com/47489629/115026195-2c2d1a80-9efd-11eb-87ef-f0cdc7e071a1.png)

Reply:
業務での使用例が書いてある記事がありましたー

https://thinkit.co.jp/article/9617